mcstatus/forge_data.py

```diff
@@ -0,0 +1,178 @@
+"""Forge Data Decoder
+
+Forge mod data is encoded into a UTF-16 string that represents
+binary data containing data like the forge mod loader network
+version, a big list of channels that all the forge mods use,
+and a list of mods the server has.
+
+For more information see this file from forge itself:
+https://github.com/MinecraftForge/MinecraftForge/blob/42115d37d6a46856e3dc914b54a1ce6d33b9872a/src/main/java/net/minecraftforge/network/ServerStatusPing.java
+"""
+
+from __future__ import annotations
+
+import io
+from dataclasses import dataclass
+from typing import Final, TYPE_CHECKING
+
+from mcstatus.protocol.connection import Connection
+
+VERSION_FLAG_IGNORE_SERVER_ONLY: Final = 0b1
+IGNORE_SERVER_ONLY: Final = "<not required for client>"
+
+
+if TYPE_CHECKING:
+    from typing_extensions import NotRequired, Self, TypedDict
+
+    class ForgeDataChannel(TypedDict):
+        res: str
+        """Channel name and ID (for example ``fml:handshake``)."""
+        version: str
+        """Channel version (for example ``1.2.3.4``)."""
+        required: bool
+        """Is this channel required for client to join?"""
+
+    class ForgeDataMod(TypedDict):
+        modid: str
+        modmarker: str
+        """Mod version."""
+
+    class RawForgeData(TypedDict):
+        fmlNetworkVersion: int  # noqa: N815 # camel case
+        channels: list[ForgeDataChannel]
+        mods: list[ForgeDataMod]
+        d: NotRequired[str]
+        truncated: NotRequired[bool]
+
+else:
+    ForgeDataChannel = dict
+    ForgeDataMod = dict
+    RawForgeData = dict
+
+
+@dataclass
+class ForgeData:
+    fml_network_version: int
+    """Forge Mod Loader network version."""
+    channels: list[ForgeDataChannel]
+    """List of channels, both for mods and non-mods."""
+    mods: list[ForgeDataMod]
+    """List of mods"""
+    truncated: bool
+    """Is the mods list and or channel list incomplete?"""
+
+    @staticmethod
+    def _decode_optimized(string: str) -> Connection:
+        """Decode buffer from UTF-16 optimized binary data `string`."""
+        text = io.StringIO(string)
+
+        def read() -> int:
+            result = text.read(1)
+            if not result:
+                return 0
+            return ord(result)
+
+        size = read() | (read() << 15)
+
+        buffer = Connection()
+        value, bits = 0, 0
+        for _ in range(len(string) - 2):
+            while bits >= 8:
+                buffer.receive((value & 0xFF).to_bytes(length=1, byteorder="big", signed=False))
+                value >>= 8
+                bits -= 8
+            value |= (read() & 0x7FFF) << bits
+            bits += 15
+
+        while buffer.remaining() < size:
+            buffer.receive((value & 0xFF).to_bytes(length=1, byteorder="big", signed=False))
+            value >>= 8
+            bits -= 8
+        return buffer
```

Comment:
Would it be possible to subclass `Connection`, overriding the `read` method to the function you have above, and use our existing logic to read the data off of here?

From what I can see, it seems that the size is read as a simple 2 byte number (short), and the code below looks like it's just reading a varint. I'm not a huge fan of this logic being essentially repeated in here, instead of making use of the already existing code for it.

That said, I didn't look too close at this, so if it is doing something special that differs from our connection implementation, let me know so we can discuss how to maybe at least make this a bit more readable.

Reply:
I simply copied the implementation forge uses, I can look into trying to do this.

Reply:
Started looking into this with 64ee784

Reply:
It does indeed end up with the effect of reading shorts, and you are correct it is very similar to reading varints, but it's not quite the same. This implementation reads chunks of 16 bits (2 bytes, short) but ignores the sign bit and ends up only using 15 bits. I tried doing a thing where it reads just one byte at a time, but with the ignore sign bit stuff it would have ended up more complicated than the implementation given. There also isn't any mechanism for segmenting the data up, the remainder data is kept in the buffer and added on to the next bytes read.

Reply:
Oh, I see, I figured it was pretty much the same code as for varints, just differently modeled.

> This implementation reads chunks of 16 bits (2 bytes, short) but ignores the sign bit and ends up only using 15 bits. I tried doing a thing where it reads just one byte at a time, but with the ignore sign bit stuff it would have ended up more complicated than the implementation given

Huh, that's a weird way to handle this, why would you ignore every 16th bit? Isn't that literally just throwing away data that could be useful? I wonder what the implementors had in mind with that. Did you find some docs or explanations, or at least references to code that explains some of how this is supposed to work exactly, and why it uses this kind of model? Especially considering it's so different to one that minecraft was already using.

> There also isn't any mechanism for segmenting the data up, the remainder data is kept in the buffer and added on to the next bytes read.

Could you expand on what you mean by this? What kind of reminder? From what I understood from your explanation above, you're supposed to be reading 2 bytes at once for every read, and then work with those 15 bits of data (dropping that 1 bit as mentioned). The rest of the data will obviously remain in a buffer, until read from again. What's the problem with segmenting the data? Is the issue just that the segments would need to be 15 bits rather than a byte? What is the "remainder" you're talking about?

---

Unless impossible, could we then create another connection-like class from zero, with the read function returning those 2 bytes, and other functions working with that? This would at least make it much easier to read through the implementations of this code, as the individual parts will be abstracted away. As an example, if the part I was talking about isn't reading varints but X, let's make `read_X`. Same for pretty much all of the operations here.

Or is there some reason this wouldn't be feasable? I'm still not entirely sure what you meant by it being unfeasable to segment the data up.

Reply:
Segment thing was about how it's 15 bits for each character and it reads 8 bits at a time, meaning it has to keep track of "remainder" bits as I called them to use for the next byte it reads.

I could definitely make some functions in the string buffer class for doing things like reading the size and could make the optimized read into a method on that class as well if that's what you would like me to do. Then the `_decode_optimized` function would only really be setting up the string buffer instance and returning the results of having it do all the reading.

Reply:
Actually, StringBuffer does handle reading a single byte at the moment. `read_short` still relies on it being a bytearray.

Reply:
Moved the rest of the reading code into StringBuffer in 534c366

Reply:
Thanks!